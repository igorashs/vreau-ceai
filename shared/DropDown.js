import styled from 'styled-components';
import { Button } from '@/shared/Button';
import BoxRemoveSvg from '@/icons/box-remove.svg';
import ArrowSvg from '@/icons/arrow.svg';
import { useState } from 'react';

const Wrapper = styled.div`
  display: grid;
  gap: var(--baseline);
  padding-bottom: calc(var(--baseline) / 2);
  border-bottom: 1px solid var(--layout);
`;

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  align-items: center;
  gap: calc(var(--baseline) / 2);
`;

const Action = styled.div`
  flex: 1;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  gap: calc(var(--baseline) / 2);

  button:last-child {
    transform: rotate(-90deg);
  }

  ${({ hide }) => hide && 'button:last-child {transform: rotate(90deg);}'}
`;

const Body = styled.div`
  ${({ hide }) => hide && 'display: none;'}
`;

export function DropDown({
  title,
  label,
  onDeleteClick,
  showInitial = false,
  customHeading,
  children
}) {
  const [hide, setHide] = useState(!showInitial);

  return (
    <Wrapper>
      <Header>
        {customHeading ? (
          customHeading
        ) : (
          <>
            <h5>{title}</h5>
            {label && <small>{label}</small>}
          </>
        )}
        <Action hide={hide}>
          {onDeleteClick && (
            <Button icon btnStyle="danger" onClick={onDeleteClick} noPadding>
              <BoxRemoveSvg />
            </Button>
          )}
          <Button icon onClick={() => setHide((e) => !e)} noPadding>
            <ArrowSvg />
          </Button>
        </Action>
      </Header>
      <Body hide={hide}>{children}</Body>
    </Wrapper>
  );
}
