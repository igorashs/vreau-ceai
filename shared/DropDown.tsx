import styled from 'styled-components';
import Button from '@/shared/Button';
import BoxRemoveSvg from '@/icons/box-remove.svg';
import ArrowSvg from '@/icons/arrow.svg';
import { useState } from 'react';

const Wrapper = styled.div`
  display: grid;
  gap: var(--baseline);
  padding-bottom: calc(var(--baseline) / 2);
  border-bottom: 1px solid var(--layout);
`;

const Header = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  align-items: center;
  gap: calc(var(--baseline) / 2);
`;

type HideStyledProps = {
  hide?: boolean;
};

const Action = styled.div<HideStyledProps>`
  flex: 1;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  gap: calc(var(--baseline) / 2);

  button:last-child {
    transform: rotate(-90deg);
  }

  ${({ hide }) => hide && 'button:last-child {transform: rotate(90deg);}'}
`;

const Body = styled.div<HideStyledProps>`
  ${({ hide }) => hide && 'display: none;'}
`;

type DropDownProps = {
  title: string;
  label?: string;
  showInitial?: boolean;
  onDeleteClick?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
  ) => void;
  customHeading?: React.ReactElement;
};

const DropDown = ({
  title,
  label,
  onDeleteClick,
  showInitial = false,
  customHeading,
  children,
}: React.PropsWithChildren<DropDownProps>) => {
  const [hide, setHide] = useState(!showInitial);

  return (
    <Wrapper>
      <Header>
        {customHeading || (
          <>
            <h5>{title}</h5>
            {label && <small>{label}</small>}
          </>
        )}
        <Action hide={hide}>
          {onDeleteClick && (
            <Button icon btnStyle="danger" onClick={onDeleteClick} noPadding>
              <BoxRemoveSvg />
            </Button>
          )}
          <Button icon onClick={() => setHide((e) => !e)} noPadding>
            <ArrowSvg />
          </Button>
        </Action>
      </Header>
      <Body hide={hide}>{children}</Body>
    </Wrapper>
  );
};

export default DropDown;
