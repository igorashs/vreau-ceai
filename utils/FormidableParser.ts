import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
