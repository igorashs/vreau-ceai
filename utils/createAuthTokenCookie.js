import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
