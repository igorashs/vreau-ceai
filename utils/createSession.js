import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
