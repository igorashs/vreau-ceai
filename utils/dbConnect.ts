import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
