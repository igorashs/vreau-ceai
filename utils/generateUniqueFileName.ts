import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
