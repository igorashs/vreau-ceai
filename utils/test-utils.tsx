import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
