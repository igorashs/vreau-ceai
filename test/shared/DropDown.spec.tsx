import { render, fireEvent } from '@/utils/test-utils';
import DropDown from '@/shared/DropDown';

describe('DropDown', () => {
  it('renders correctly', () => {
    const title = 'DropDown title';

    const { container, getByText, getByRole } = render(
      <DropDown title={title}>content</DropDown>,
    );

    expect(getByText(title)).toBeVisible();
    expect(getByText('content')).not.toBeVisible();
    expect(getByText('content')).toHaveStyleRule('display', 'none');
    expect(getByRole('button').parentElement).toHaveStyleRule(
      'transform',
      'rotate(90deg)',
      {
        modifier: 'button:last-child',
      },
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('.onToggleHideClick', () => {
    it('toggles content', () => {
      const title = 'DropDown title';

      const { getByText, getByRole } = render(
        <DropDown title={title}>content</DropDown>,
      );

      const toggleBtn = getByRole('button');
      fireEvent.click(toggleBtn);

      expect(getByText('content')).toBeVisible();
      expect(toggleBtn.parentElement).toHaveStyleRule(
        'transform',
        'rotate(-90deg)',
        {
          modifier: 'button:last-child',
        },
      );

      fireEvent.click(toggleBtn);
      expect(getByText('content')).not.toBeVisible();
    });
  });

  describe('.onDeleteClick', () => {
    it('renders delete btn', () => {
      const title = 'DropDown title';
      const onDeleteClickMock = jest.fn();

      const { getAllByRole } = render(
        <DropDown title={title} onDeleteClick={onDeleteClickMock}>
          content
        </DropDown>,
      );

      const buttons = getAllByRole('button');

      expect(buttons.length).toBe(2);
      fireEvent.click(buttons[0]);

      expect(onDeleteClickMock).toBeCalledTimes(1);
    });
  });

  describe('label prop', () => {
    it('renders with label', () => {
      const title = 'DropDown title';
      const label = 'DropDown label';

      const { getByText } = render(
        <DropDown title={title} label={label}>
          content
        </DropDown>,
      );

      expect(getByText(label)).toBeVisible();
    });
  });

  describe('CustomHeading prop', () => {
    it('renders with a custom heading', () => {
      const title = 'DropDown title';
      const CustomHeading = () => <h5>custom heading</h5>;

      const { getByText, queryByText } = render(
        <DropDown title={title} CustomHeading={CustomHeading}>
          content
        </DropDown>,
      );

      expect(queryByText(title)).not.toBeInTheDocument();
      expect(getByText('custom heading')).toBeVisible();
    });
  });
});
