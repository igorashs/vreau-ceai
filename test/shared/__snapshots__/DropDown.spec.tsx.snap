// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown renders correctly 1`] = `
.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: 4px;
  box-shadow: none;
  padding: 0;
}

.c3:hover {
  background-color: var(--accent);
}

.c0 {
  display: grid;
  gap: var(--baseline);
  padding-bottom: calc(var(--baseline) / 2);
  border-bottom: 1px solid var(--layout);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: calc(var(--baseline) / 2);
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: calc(var(--baseline) / 2);
}

.c2 button:last-child {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.c2 button:last-child {
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c4 {
  display: none;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <h5>
      DropDown title
    </h5>
    <div
      class="c2"
    >
      <button
        class="c3"
      >
        <svg />
      </button>
    </div>
  </div>
  <div
    class="c4"
  >
    content
  </div>
</div>
`;
